File: frontend/src/types/artifact.ts
export type ArtifactStatusState = "pending" | "rendering" | "done" | "failed";

export type ArtifactStatus = {
  state: ArtifactStatusState;
  approved: boolean;
  error: string | null;
};

export type ArtifactMetaV1 = {
  schema_version: 1;
  render_id: string;
  preset_id: string;
  preset_version: string;
  seed: number;
  params: Record<string, unknown>;
  song_id: string;
  analysis_id: string;
  fps: number;
  duration: number;
  frame_count: number;
  status: ArtifactStatus;
};

export function isCompatibleArtifact(raw: unknown): raw is ArtifactMetaV1 {
  if (typeof raw !== "object" || raw === null) return false;
  return (raw as Record<string, unknown>)["schema_version"] === 1;
}

File: backend/render/artifact.py
import hashlib
import json
from typing import Any, Literal

from pydantic import BaseModel, Field


class StatusBlock(BaseModel):
    state: Literal["pending", "rendering", "done", "failed"] = "pending"
    approved: bool = False
    error: str | None = None


class ArtifactMetaV1(BaseModel):
    schema_version: Literal[1] = 1
    render_id: str
    preset_id: str
    preset_version: str
    seed: int
    params: dict[str, Any]
    song_id: str
    analysis_id: str
    fps: float
    duration: float
    frame_count: int
    status: StatusBlock = Field(default_factory=StatusBlock)


def make_render_id(
    song_id: str,
    preset_id: str,
    seed: int,
    params: dict[str, Any],
) -> str:
    payload = json.dumps(
        {"song_id": song_id, "preset_id": preset_id, "seed": seed, "params": params},
        sort_keys=True,
    ).encode()
    return hashlib.sha256(payload).hexdigest()[:16]
